Add Jakarta Data documentation, dependency management and tests (#47012)
quarkusio/quarkus: docs/src/main/asciidoc/hibernate-orm.adoc

Having multiple JSON (or XML) format mappers registered for the same persistence unit will result in an exception, because of the ambiguity.
====

[[jakarta-data]]
== Static metamodel and Jakarta Data

> Both static metamodel and Jakarta Data capabilities of Hibernate ORM are available to the users

Reviewer: ```suggestion
Both static metamodel and Jakarta Data capabilities of Hibernate ORM are available in Quarkus
```

through the `hibernate-jpamodelgen` annotation processor. Since it is an annotation processor,

> users are required to make an extra step of configuring the annotation processor in their build tool of choice:

Reviewer: Don't ask me why, or where the rule is written, but in Quarkus documentation we're supposed to address directly to the reader.

```suggestion
you must configure the annotation processor in your build tool:
```

                <!-- Note, no artifact version is required, it's managed by Quarkus.  -->
            </path>
            <!-- other processors that may be required by your app -->
        </annotationProcessorPaths>
        <!-- Other compiler plugin configuration options -->
    </configuration>
</plugin>
----

[source,gradle,role="secondary asciidoc-tabs-target-sync-gradle"]
.build.gradle
----
// Enforce the version management of your annotation processor dependencies,
// so that there's no need to define an explicit version of the hibernate-jpamodelgen
annotationProcessor enforcedPlatform("${quarkusPlatformGroupId}:${quarkusPlatformArtifactId}:${quarkusPlatformVersion}")
annotationProcessor 'org.hibernate.orm:hibernate-jpamodelgen'
----

=== Static metamodel

The generated static metamodel allows for building queries in a type-safe manner.
Let's consider having a simple entity:

[source,java]
----
@Entity
public class MyEntity {
    @Id
    @GeneratedValue
    public Integer id;

    @Column(unique = true)
    public String name;
}
----

A query created with the help of static metamodel may look as:

[source,java]
----
var builder = session.getCriteriaBuilder();

> var query = builder.createQuery(MyEntity.class);
> var e = query.from(MyEntity_.class);
> query.where(e.get(MyEntity_.name).equalTo(name));
> session.createQuery(query)....

Reviewer: ```suggestion
var criteria = builder.createQuery(MyEntity.class);
var e = criteria.from(MyEntity_.class);
criteria.where(e.get(MyEntity_.name).equalTo(name));
var query = session.createQuery(criteria);
var result = query.list();
```

var query = builder.createQuery(MyEntity.class);
var e = query.from(MyEntity_.class_);
query.where(e.get(MyEntity_.name).equalTo(name));
session.createQuery(query)....
----

For a more detailed overview of static metamodel, please refer to the link:{jakarta-persistence-spec-url}#a6072[Jakarta Persistence specification]

=== Jakarta Data

Jakarta Data requires, besides having the `hibernate-jpamodelgen` annotation processor in place, one extra dependency to be added:

[source,xml,role="primary asciidoc-tabs-target-sync-cli asciidoc-tabs-target-sync-maven"]
.pom.xml
----
<dependency>
    <groupId>jakarta.data</groupId>
    <artifactId>jakarta.data-api</artifactId>
</dependency>
----

[source,gradle,role="secondary asciidoc-tabs-target-sync-gradle"]
.build.gradle
----
implementation 'jakarta.data:jakarta.data-api'
----

With this dependency, and the annotation processor in place you could simply create your repositories as follows:

[source,java]
----
@Repository
public interface MyRepository {

    @Find
    Optional<MyEntity> findById(Integer id);

    @Query("select e from MyEntity e where e.name like :name")
    List<MyEntity> findByName(String name);

>     @Insert

Author: hey 😃 👋🏻 
thanks for taking a look. yeah .. wasn't sure if we want to use one of those interfaces or just show that we can write our own stuff. But if you think `CRUDRepository` would be better, I'll update it 😃 👍🏻

>     @Insert

Reviewer 2: I think it would, the methods are just boilerplate here.

>     @Insert

Author: ok, added some callouts to this example and mentioned available data interfaces there.
